gym_roboy/envs/ros_proxy.py

```diff
@@ -49,38 +57,87 @@ def __init__(self, timeout_secs: int = 2):
-        self._step_size = 0.1
+        self.goal_client = self.node.create_client(GymGoal, 'gym_goal')
+
+        self.sphere_axis0 = self.node.create_publisher(msg_type=Float32, topic="/sphere_axis0/sphere_axis0/target")
+        self.sphere_axis1 = self.node.create_publisher(msg_type=Float32, topic="/sphere_axis1/sphere_axis1/target")
+        self.sphere_axis2 = self.node.create_publisher(msg_type=Float32, topic="/sphere_axis2/sphere_axis2/target")
+
+    def _log_robot_state(self, robot_state):
+        q_pos = robot_state.q
+        q_vel = robot_state.qdot
+        qpos_str = str(q_pos).strip('[]')
+        qvel_str = str(q_vel).strip('[]')
+        self.node.get_logger().info("joint angles: %s" % qpos_str)
+        self.node.get_logger().info("joint velocity: %s" % qvel_str)
 
     def forward_reset_command(self):
-        request = GymStep.Request()
-        request.step_size = self._step_size
-        future = self.reset_client.call_async(request)
-        self._wait_until_future_complete_or_timeout(future)
-        return self._make_robot_state(service_response=future.result())
+        while self._wait_until_future_complete_or_timeout(self.reset_client):
+            request = GymStep.Request()
+            request.step_size = self._step_size
+            future = self.reset_client.call_async(request)
+            rclpy.spin_until_future_complete(self.node, future)
+            return self._make_robot_state(service_response=future.result())
 
     @staticmethod
     def _make_robot_state(service_response) -> MsjRobotState:
-        return MsjRobotState(joint_angle=service_response.q[3:],
-                             joint_vel=service_response.qdot[3:])
+        return MsjRobotState(joint_angle=service_response.q,
+                             joint_vel=service_response.qdot)
 
     def forward_step_command(self, action):
-        req = GymStep.Request()
-        req.set_points = action
-        req.step_size = self._step_size
-        future = self.step_client.call_async(req)
-        self._wait_until_future_complete_or_timeout(future)
-        return self._make_robot_state(future.result())
-
-    def _wait_until_future_complete_or_timeout(self, future):
-        if not self.step_client.wait_for_service(timeout_sec=self._timeout_secs):
+        while self._wait_until_future_complete_or_timeout(self.step_client):
+            req = GymStep.Request()
+            req.set_points = action
+            req.step_size = self._step_size
+            future = self.step_client.call_async(req)
+            rclpy.spin_until_future_complete(self.node, future)
+            res = future.result()
+            if res is not None:
+                self._log_robot_state(res)
+                if not res.feasible:
+                    return self.forward_reset_command()
+                return self._make_robot_state(res)
+
+
+    def _wait_until_future_complete_or_timeout(self, client):
```

Comment:
Now we have a new issue. The `_wait_until_future_complete_or_timeout()` method does not wait for any future to be completed. Also, it is used in a while loop that is not necessary, because it either is available in x secs or it throws an Error. Waht about we get rid of the while loops, the return True, and leave a void function that checks the service or timeouts.
```python
def _check_service_is_available_or_timeout(self, service_client) -> None:
    ...
```

Reply:
Since the `TimeoutError` is shared by all methods: `forward_step_command(), forward_reset_command(), get_new_goal_joint_angle()` and `read_state()`, we should encapsulate it in a single function to avoid code duplication.